akka-projection-grpc/src/main/scala/akka/projection/grpc/internal/EventsBySlicesFirehose.scala

```diff
@@ -0,0 +1,246 @@
+        }
+        .via(catchupKillSwitch.flow)
+
+    val consumerKillSwitch = KillSwitches.shared("consumerKillSwitch")
+
+    firehose.firehoseHub
+      .map { env =>
+        firehose.updateConsumerTracking(consumerId, timestampOffset(env).timestamp, consumerKillSwitch)
+        FirehoseEventEnvelope(env)
+      }
+      .merge(catchup)
+      .via(MergeSelector.flow(consumerId, catchupKillSwitch))
+      .map(_.asInstanceOf[EventEnvelope[Event]])
+      .via(consumerKillSwitch.flow)
+  }
+
+  // can be overridden in tests
+  @nowarn("msg=never used") // firehose param used in tests
+  protected def eventsBySlices[Event](
+      entityType: String,
+      minSlice: Int,
+      maxSlice: Int,
+      offset: Offset,
+      firehose: Boolean): Source[EventEnvelope[Event], NotUsed] = {
+    val queryPluginId = "akka.persistence.r2dbc.query"
+    PersistenceQuery(system)
+      .readJournalFor[EventsBySliceQuery](queryPluginId)
+      .eventsBySlices(entityType, minSlice, maxSlice, offset)
+  }
+
+}
+
+// FIXME maybe write as GraphStage instead
+object MergeSelector {
+  import EventsBySlicesFirehose._
+  private sealed trait Mode
+  private case object CatchUpOnly extends Mode
+  private final case class Both(caughtUpTimestamp: Instant) extends Mode
+  private case object FirehoseOnly extends Mode
+
+  private val log = LoggerFactory.getLogger(classOf[MergeSelector.type])
+
+  def flow(consumerId: String, catchupKillSwitch: KillSwitch): Flow[WrappedEventEnvelope, EventEnvelope[Any], NotUsed] =
+    Flow[WrappedEventEnvelope].statefulMapConcat {
+      var mode: Mode = CatchUpOnly
+      var firehoseOffset: TimestampOffset = TimestampOffset(Instant.EPOCH, Map.empty)
+
+      def updateFirehoseOffset(env: EventEnvelope[Any]): Unit = {
+        val offset = timestampOffset(env)
+        if (offset.timestamp.isAfter(firehoseOffset.timestamp))
+          firehoseOffset = offset // update when newer
+      }
+
+      def isCaughtUp(env: EventEnvelope[Any]): Boolean = {
+        val offset = timestampOffset(env)
+        firehoseOffset.timestamp != Instant.EPOCH && !firehoseOffset.timestamp.isAfter(offset.timestamp)
+      }
+
+      () => {
+        case FirehoseEventEnvelope(env) =>
+          mode match {
+            case CatchUpOnly =>
+              updateFirehoseOffset(env)
+              Nil // skip
+            case Both(_) =>
+              updateFirehoseOffset(env)
+              env :: Nil // emit
+            case FirehoseOnly =>
+              env :: Nil // emit
+          }
+
+        case CatchupEventEnvelope(env) =>
+          mode match {
+            case CatchUpOnly =>
+              if (isCaughtUp(env)) {
+                val timestamp = timestampOffset(env).timestamp
+                log.debug("Consumer [{}] caught up at [{}]", consumerId, timestamp)
+                mode = Both(timestamp)
+              }
+              env :: Nil // emit
+            case Both(caughtUpTimestamp) =>
```

Comment:
I don't quite understand the need for the both mode, if we caught up, can't we switch over to firehose and cancel the catchup?

Reply:
My thought was that for similar reason as we need the backtracking we can't be sure we have seen all events from one query just because the other query included them. We might get away with it if we let the backtracking cover that scenario as well, but missing event and then catch it via backtracking introduce long latency for that event. Then we should be sure that it's very rare.